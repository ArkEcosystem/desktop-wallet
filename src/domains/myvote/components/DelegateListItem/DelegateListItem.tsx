import { Button } from "app/components/Button";
import { Icon } from "app/components/Icon";
import { Avatar } from "domains/wallet/components/Avatar";
import React from "react";
import { useTranslation } from "react-i18next";

type DelegateListItemProps = {
	delegateAddress?: string;
	delegateName?: string;
	rank?: number;
	votes?: number;
	msqUrl?: string;
	commissionPercentage?: number;
	commissionDaily?: number;
	payout?: string;
	min?: number;
	selected?: any[];
	onSelect?: ({ address, username, rank }: { address: string; username: string; rank: number }) => void;
};

export const DelegateListItem = ({
	delegateAddress,
	delegateName,
	rank,
	votes,
	msqUrl,
	commissionPercentage,
	commissionDaily,
	payout,
	min,
	selected,
	onSelect,
}: DelegateListItemProps) => {
	const { t } = useTranslation();
	const isSelected = selected?.find((selectedDelegate: any) => selectedDelegate.username === delegateName) || false;

	return (
		<tr className="border-b border-theme-neutral-200">
			<td className="py-5">
				<Avatar address={delegateAddress!} />
			</td>

			<td className="py-5 font-bold">{delegateName}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">#{rank}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">{votes}%</td>

			<td className="py-5">
				{msqUrl && (
					<div className="flex justify-center h-full">
						<a href={msqUrl} target="_blank" rel="noreferrer">
							<Icon name="Msq" className="text-xl text-theme-primary" />
						</a>
					</div>
				)}
			</td>

			<td className="py-5 font-bold text-theme-neutral-dark">
				{commissionPercentage && <span>{commissionPercentage}%</span>}
			</td>

			<td className="py-5 font-bold text-theme-neutral-dark">{payout && <span>{payout}</span>}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">{min && <span>{min} Ѧ</span>}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">
				{commissionDaily && <span>{commissionDaily} Ѧ</span>}
			</td>

			<td className="py-5">
				<div className="text-right">
					<Button
						variant="plain"
						color={isSelected ? "danger" : "primary"}
						onClick={() => onSelect?.({ address: delegateAddress!, username: delegateName!, rank: rank! })}
						data-testid="DelegateListItem__button--toggle"
					>
						{isSelected ? t("COMMON.UNSELECT") : t("COMMON.SELECT")}
					</Button>
				</div>
			</td>
		</tr>
	);
};

DelegateListItem.defaultProps = {
	selected: [],
};
