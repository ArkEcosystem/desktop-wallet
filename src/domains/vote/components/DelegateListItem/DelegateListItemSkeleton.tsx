import { Circle } from "app/components/Circle";
import React from "react";
import Skeleton from "react-loading-skeleton";

export const DelegateListItemSkeleton = () => (
	<tr data-testid="DelegateListItemSkeleton" className="border-b border-dotted border-theme-neutral-300">
		<td className="py-5">
			<Circle aria-label="network" className="border-theme-neutral-200" />
		</td>

		<td className="py-5">
			<Skeleton width="80%" height={6} className="mt-4" />
		</td>

		<td className="py-5">
			<Skeleton height={6} width="40%" className="mt-4" />
		</td>

		<td className="py-5 font-bold text-theme-neutral-dark">
			<Skeleton height={6} width="40%" className="mt-4" />
		</td>

		<td className="py-5">
			<Skeleton width={24} height={24} />
		</td>

		<td className="py-5 font-bold text-theme-neutral-dark">
			<Skeleton width="80%" height={6} className="mt-4" />
		</td>

		<td className="py-5 font-bold text-theme-neutral-dark">
			<Skeleton width="80%" height={6} className="mt-4" />
		</td>

		<td className="py-5 font-bold text-theme-neutral-dark">
			<Skeleton width="80%" height={6} className="mt-4" />
		</td>

		<td className="py-5 font-bold text-theme-neutral-dark">
			<Skeleton height={6} width="60%" className="mt-4" />
		</td>

		<td className="py-5">
			<Skeleton width={100} height={40} />
		</td>
	</tr>
);
