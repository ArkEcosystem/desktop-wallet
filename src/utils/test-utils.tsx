import { render } from "@testing-library/react";
import { i18n } from "app/i18n";
import React from "react";
import { I18nextProvider } from "react-i18next";

const WithProviders = ({ children }: { children: React.ReactNode }) => {
	return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};

const customRender = (component: React.ReactElement, options: any) =>
	render(component, { wrapper: WithProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
