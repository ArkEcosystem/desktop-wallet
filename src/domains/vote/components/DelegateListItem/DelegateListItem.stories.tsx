import { Table } from "app/components/Table";
import React from "react";

import { DelegateListItem } from "./DelegateListItem";

export default { title: "Domains / Vote / Components / DelegateListItem" };

const data = [
	{
		delegateAddress: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT",
		delegateName: "Delegate 1",
		rank: 1,
		votes: 3.43,
		msqUrl: "https://marketsquare.ark.io",
		commissionPercentage: 80,
		commissionDaily: 2.387082496,
		payout: "Daily",
		min: 500,
	},
];

const columns = [
	{
		Header: "",
		accessor: "delegateAddressAvatar",
		disableSortBy: true,
	},
	{
		Header: "Delegate Name",
		accessor: "delegateName",
	},
	{
		Header: "Rank",
		accessor: "rank",
	},
	{
		Header: "Votes",
		accessor: "votes",
	},
	{
		Header: "Profile",
		accessor: "profile",
		disableSortBy: true,
		className: "flex justify-center",
	},
	{
		Header: "Comm.",
		accessor: "commissionPercentage",
	},
	{
		Header: "Payout",
		accessor: "payout",
	},
	{
		Header: "Min",
		accessor: "min",
	},
	{
		Header: "Commission (Daily)",
		accessor: "commissionDaily",
	},
	{
		Header: "",
		accessor: "onSelect",
		disableSortBy: true,
	},
];

export const Default = () => (
	<Table columns={columns} data={data}>
		{(rowData: any) => <DelegateListItem {...rowData} />}
	</Table>
);
