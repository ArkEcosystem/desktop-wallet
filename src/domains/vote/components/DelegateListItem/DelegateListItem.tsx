import { Contracts } from "@arkecosystem/platform-sdk";
import { Avatar } from "app/components/Avatar";
import { Button } from "app/components/Button";
import React from "react";
import { useTranslation } from "react-i18next";

import { DelegateListItemSkeleton } from "./DelegateListItemSkeleton";

type DelegateListItemProps = {
	index: number;
	delegate: Contracts.WalletData;
	selected?: any[];
	isLoading?: boolean;
	onSelect?: ({ address, username, rank }: { address: string; username: string; rank: number }) => void;
};

export const DelegateListItem = ({ index, delegate, selected, isLoading, onSelect }: DelegateListItemProps) => {
	const { t } = useTranslation();
	const isSelected =
		selected?.find((selectedDelegate: any) => selectedDelegate.username === delegate.username()) || false;

	if (isLoading) {
		return <DelegateListItemSkeleton />;
	}

	return (
		<tr className={`border-b border-dotted border-theme-neutral-300 ${isSelected && "bg-theme-success-contrast"}`}>
			<td className="py-5">
				<Avatar address={delegate.address()} noShadow />
			</td>

			<td className="py-5 font-bold">{delegate.username()}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">#{delegate.rank()}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">%</td>

			{/* <td className="py-5">
				{msqUrl && (
					<div className="flex justify-center h-full">
						<a href={msqUrl} target="_blank" rel="noopener noreferrer">
							<Icon name="Msq" className="text-xl text-theme-primary" />
						</a>
					</div>
				)}
			</td>

			<td className="py-5 font-bold text-theme-neutral-dark">
				{commissionPercentage && <span>{commissionPercentage}%</span>}
			</td>

			<td className="py-5 font-bold text-theme-neutral-dark">{payout && <span>{payout}</span>}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">{min && <span>{min} Ѧ</span>}</td>

			<td className="py-5 font-bold text-theme-neutral-dark">
				{commissionDaily && <span>{commissionDaily} Ѧ</span>}
			</td> */}

			<td className="py-5">
				<div className="text-right">
					<Button
						variant="plain"
						color={isSelected ? "danger" : "primary"}
						onClick={() =>
							onSelect?.({
								address: delegate.address(),
								username: delegate.username()!,
								rank: delegate.rank()!,
							})
						}
						data-testid={`DelegateListItem__toggle-${index}`}
					>
						{isSelected ? t("COMMON.UNSELECT") : t("COMMON.SELECT")}
					</Button>
				</div>
			</td>
		</tr>
	);
};

DelegateListItem.defaultProps = {
	selected: [],
	isLoading: false,
};
