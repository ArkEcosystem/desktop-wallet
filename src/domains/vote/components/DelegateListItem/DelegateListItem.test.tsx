import React from "react";
import { act, fireEvent, render } from "test-utils";

import { DelegateListItem } from "./DelegateListItem";

const data = {
	delegateAddress: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT",
	delegateName: "Delegate 1",
	rank: 1,
	votes: 3.43,
	msqUrl: "https://marketsquare.ark.io",
	commissionPercentage: 80,
	commissionDaily: 2.387082496,
	payout: "Daily",
	min: 500,
};

describe("DelegateListItem", () => {
	it("should render", () => {
		const { container, asFragment } = render(
			<table>
				<tbody>
					<DelegateListItem {...data} />
				</tbody>
			</table>,
		);

		expect(container).toBeTruthy();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should emit action on select button", () => {
		const onSelect = jest.fn();
		const { container, asFragment, getByTestId } = render(
			<table>
				<tbody>
					<DelegateListItem {...data} onSelect={onSelect} />
				</tbody>
			</table>,
		);
		const selectButton = getByTestId("DelegateListItem__button--toggle");

		act(() => {
			fireEvent.click(selectButton);
		});

		expect(container).toBeTruthy();
		expect(onSelect).toHaveBeenCalledWith({
			address: data.delegateAddress,
			username: data.delegateName,
			rank: data.rank,
		});
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render the selected delegate", () => {
		const selected = [{ address: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT", username: "Delegate 1", rank: 1 }];
		const { container, asFragment, getByTestId } = render(
			<table>
				<tbody>
					<DelegateListItem {...data} selected={selected} />
				</tbody>
			</table>,
		);

		expect(container).toBeTruthy();
		expect(getByTestId("DelegateListItem__button--toggle")).toHaveTextContent("Unselect");
		expect(asFragment()).toMatchSnapshot();
	});
});
