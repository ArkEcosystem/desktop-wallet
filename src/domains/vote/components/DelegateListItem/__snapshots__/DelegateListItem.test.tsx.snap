// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelegateListItem should emit action on select button 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="border-b border-theme-neutral-200 false"
      >
        <td
          class="py-5"
        >
          <div
            class="Avatar__AvatarWrapper-sc-1vsy2s2-0 fLVAbv"
            data-testid="Avatar"
          >
            <div
              class="overflow-hidden rounded-full"
            >
              <img
                alt="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
                src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\"><style>circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(244, 67, 54)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"45\\" cx=\\"60\\" cy=\\"30\\" fill=\\"rgb(3, 169, 244)\\"/><circle r=\\"35\\" cx=\\"10\\" cy=\\"40\\" fill=\\"rgb(205, 220, 57)\\"/><circle r=\\"50\\" cx=\\"40\\" cy=\\"50\\" fill=\\"rgb(156, 39, 176)\\"/></svg>"
                title="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
              />
            </div>
          </div>
        </td>
        <td
          class="py-5 font-bold"
        >
          Delegate 1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          #1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          3.43%
        </td>
        <td
          class="py-5"
        >
          <div
            class="flex justify-center h-full"
          >
            <a
              href="https://marketsquare.ark.io"
              rel="noreferrer"
              target="_blank"
            >
              <div
                class="sc-AxirZ iQaTP text-xl text-theme-primary"
                height="1em"
                width="1em"
              >
                <svg>
                  msq.svg
                </svg>
              </div>
            </a>
          </div>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            80%
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            Daily
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            500 Ѧ
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            2.387082496 Ѧ
          </span>
        </td>
        <td
          class="py-5"
        >
          <div
            class="text-right"
          >
            <button
              class="sc-AxjAm ddJBbC"
              color="primary"
              data-testid="DelegateListItem__button--toggle"
              type="button"
            >
              Select
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DelegateListItem should render 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="border-b border-theme-neutral-200 false"
      >
        <td
          class="py-5"
        >
          <div
            class="Avatar__AvatarWrapper-sc-1vsy2s2-0 fLVAbv"
            data-testid="Avatar"
          >
            <div
              class="overflow-hidden rounded-full"
            >
              <img
                alt="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
                src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\"><style>circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(244, 67, 54)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"45\\" cx=\\"60\\" cy=\\"30\\" fill=\\"rgb(3, 169, 244)\\"/><circle r=\\"35\\" cx=\\"10\\" cy=\\"40\\" fill=\\"rgb(205, 220, 57)\\"/><circle r=\\"50\\" cx=\\"40\\" cy=\\"50\\" fill=\\"rgb(156, 39, 176)\\"/></svg>"
                title="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
              />
            </div>
          </div>
        </td>
        <td
          class="py-5 font-bold"
        >
          Delegate 1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          #1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          3.43%
        </td>
        <td
          class="py-5"
        >
          <div
            class="flex justify-center h-full"
          >
            <a
              href="https://marketsquare.ark.io"
              rel="noreferrer"
              target="_blank"
            >
              <div
                class="sc-AxirZ iQaTP text-xl text-theme-primary"
                height="1em"
                width="1em"
              >
                <svg>
                  msq.svg
                </svg>
              </div>
            </a>
          </div>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            80%
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            Daily
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            500 Ѧ
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            2.387082496 Ѧ
          </span>
        </td>
        <td
          class="py-5"
        >
          <div
            class="text-right"
          >
            <button
              class="sc-AxjAm ddJBbC"
              color="primary"
              data-testid="DelegateListItem__button--toggle"
              type="button"
            >
              Select
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DelegateListItem should render the selected delegate 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="border-b border-theme-neutral-200 bg-theme-success-100"
      >
        <td
          class="py-5"
        >
          <div
            class="Avatar__AvatarWrapper-sc-1vsy2s2-0 fLVAbv"
            data-testid="Avatar"
          >
            <div
              class="overflow-hidden rounded-full"
            >
              <img
                alt="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
                src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\"><style>circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(244, 67, 54)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"45\\" cx=\\"60\\" cy=\\"30\\" fill=\\"rgb(3, 169, 244)\\"/><circle r=\\"35\\" cx=\\"10\\" cy=\\"40\\" fill=\\"rgb(205, 220, 57)\\"/><circle r=\\"50\\" cx=\\"40\\" cy=\\"50\\" fill=\\"rgb(156, 39, 176)\\"/></svg>"
                title="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT"
              />
            </div>
          </div>
        </td>
        <td
          class="py-5 font-bold"
        >
          Delegate 1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          #1
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          3.43%
        </td>
        <td
          class="py-5"
        >
          <div
            class="flex justify-center h-full"
          >
            <a
              href="https://marketsquare.ark.io"
              rel="noreferrer"
              target="_blank"
            >
              <div
                class="sc-AxirZ iQaTP text-xl text-theme-primary"
                height="1em"
                width="1em"
              >
                <svg>
                  msq.svg
                </svg>
              </div>
            </a>
          </div>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            80%
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            Daily
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            500 Ѧ
          </span>
        </td>
        <td
          class="py-5 font-bold text-theme-neutral-dark"
        >
          <span>
            2.387082496 Ѧ
          </span>
        </td>
        <td
          class="py-5"
        >
          <div
            class="text-right"
          >
            <button
              class="sc-AxjAm bxcsTZ"
              color="danger"
              data-testid="DelegateListItem__button--toggle"
              type="button"
            >
              Unselect
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
